### PR TITLE
Enforce cgroups limits

### LINKED ISSUE #1316: CGroups
Today cgroups are only enabled for Ubuntu. We need to enable them by default, and enforce limits.

### PR DESCRIPTION
**Pull request by hglkrijger**
## Description

- closes #1316 
- add a global configuration flag for enforcing cgroups limits
- enforce limits by default on non-systemd-managed cgroups
- exclude CustomScript and RunCommand extensions
- add telemetry for enforced limits
- adds unit tests

---

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes in the description.
- [x] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -452,6 +453,14 @@ If set, the agent will use this proxy server to access the internet. These value
 *will* override the `http_proxy` or `https_proxy` environment variables. Lastly,
 `HttpProxy.Host` is required (if to be used) and `HttpProxy.Port` is optional.
 
+#### __CGroups.EnforceLimits__
+
+_Type: Boolean_  
+_Default: y_
+
+If set, the agent will attempt to set cgroups limits for cpu and memory for the agent process itself
+as well as extension processes. See the wiki for further details on this.
```
Comment (boumenot) [inline README.md]: `...See the wiki...` - I look forward to the verbiage.

File: config/bigip/waagent.conf [hunk 2]
```diff
@@ -97,3 +97,6 @@ AutoUpdate.Enabled=y
 
 # Add firewall rules to protect access to Azure host node services
 OS.EnableFirewall=y
+
+# Enforce control groups limits on the agent and extensions
+CGroups.EnforceLimits=y
```
Comment (boumenot) [inline config/bigip/waagent.conf]: Does BigIP have CGroups?

I guess it doesn't really matter too much.  If you can do this, you'll log a warning and carry on as usual.

File: README.md (hunk 1 above)
Comment (hglkrijger) [reply README.md]: :+1:

File: config/bigip/waagent.conf (hunk 2 above)
Comment (hglkrijger) [reply config/bigip/waagent.conf]: Figured it was better to assume it does.

File: azurelinuxagent/common/cgroups.py [hunk 3]
```diff
@@ -30,6 +32,15 @@
 WRAPPER_CGROUP_NAME = "Agent+Extensions"
 METRIC_HIERARCHIES = ['cpu', 'memory']
 MEMORY_DEFAULT = -1
+UNLIMITED_INSTANCES = ['customscript', 'runcommand']
```
Comment (jasonzio) [inline azurelinuxagent/common/cgroups.py]: Consider putting this list of "extensions which are exceptions" into the waagent.conf file. This allows customers to make exceptions for extensions we don't know about today without requiring an agent re-release.

Comment (hglkrijger) [reply azurelinuxagent/common/cgroups.py]: Good call. I will open a new PR for that change so I can get some coverage on this one overnight.

**Pull request merged**